# <FONT COLOR="blue"> **Samsung Innovation Campus**</FONT>  

## <FONT COLOR="blue">**Codificación y programación**</FONT>

# **Capítulo 1. Conceptos básicos de programación e iniciando Python**

Descripción del capítulo

## Objetivos del capítulo

Los estudiantes podrán configurar un ambiente de práctica para Python y aprender los fundamentos de programación. Los estudiantes se familiarizarán con los fundamentos de Python como entradas/salidas, tipos de datos, y variables. Adicionalmente, adquirirán habilidades para la resolución de problemas de codificación a través de varios operadores, condicionales y ciclos.

## Contenido del capítulo

* Unidad 1. Programación secuencial
* Unidad 2. Planeación para la programación
* Unidad 3. Fundamentos de tipos de datos numéricos y operaciones aritméticas
* Unidad 4. Variables y entradas
* Unidad 5. Operaciones lógicas y de comparación
* Unidad 6. Sentencias condicionales-1: Condiciones y toma de decisiones
* Unidad 7. Sentencias condicionales-2: Toma de decisiones en dos direcciones y aplicación de sentencias condicionales
* Unidad 8. Bucle-1
* Unidad 9. Bucle-2

# **Unidad 3. Tipos de Datos numéricos básicos y operaciones aritméticas**

## Objetivos de aprendizaje

* Ser capaz de realizar sumas, restas, multiplicaciones divisiones, y potenciaciones usando Python. De igual forma entender y estar en capacidad de usar funciones avanzadas de Python para encontrar el cociente y el residuo de una división.
* Entender que hay diferentes tipos de datos en el procesamiento de los mismos y estar en capacidad de ejecutar cálculos numéricos usando relaciones entre tipos de datos y operadores.
* Aprender cómo tratar con cadenas de caracteres (strings) y estar en capacidad de imprimir valores de datos de un código.
* Ser capaz de escribir ecuaciones usando la precedencia de operadores.

## Resumen del Aprendizaje

* Ejecutar operaciones mateméticas incluyendo operaciones aritméticas, potenciación, etc., usando tipos de datos numéricos.
* Determinar el tipo de datos de objetos y aprender cómo funcionan los operadores dependiendo del tipo de datos.
* Generar una salida o resultado usando tipos de datos string (cadena de caracteres).

## Conceptos que necesitará saber de unidades anteriores

* Imprimir una salida o resultado en la pantalla usando la función print y obtener valores ingresados por el usuario usando input.
* Cómo correr Python usando Jupyter Notebook. Python es una interfaz de Python para codificación interactiva en Python y debería poder escribir un código en Python in la condición.

## Palabras Clave

![Image](https://ambg0231.github.io/unidad3_/Keywords.png)

## <FONT COLOR="blue"> Misión </FONT>

### <FONT COLOR="blue">1. Problema del mundo real </FONT>

#### **1.1. Observar la sombra de un agujero negro**


<table style='font-size:175%; width:100%'>
  <tr>
      <td>
        <img style="float: left;"<img src="https://ambg0231.github.io/unidad3_/BlackHole.png"  width= "800px" />
      </td>
      <td>
        <ul>
            <li>La figura de abajo es la primera fotografía de un agujero negro observado por la humanidad.</li>
            <li>El proyecto fue llamado EHT (Event Horizon Telescope) y científicos de 32 países diferentes participaron.</li>
            <li>Observar un agujero negro que está muy lejos es difícil.</li>
            <li>Para resolver este problema, los miembros del proyecto EHT decidieron usar el planeta tierra entero como un gran telescopio, observar un punto en el universo usando radiotelescopios en diferentes regiones al mismo tiempo, y analizar dichas imágenes observadas.</li>
        </ul>
      </td>
  </tr>
</table>

#### **1.2. Fórmulas del área y la circunferencia de un círculo**


<table style='font-size:150%; width:100%'>
  <tr>
      <td>
          <img style="float: left;"<img src="https://ambg0231.github.io/unidad3_/Circumf.png"  width= "800px" />
      </td>
      <td>
        <ul>
            <li>Trataremos de resolver problemas matemáticos básicos de entre las varias técnicas usadas por los ingenieros y científicos en el proyecto EHT para resolver problemas.</li>
            <li>Es decir, si hay un círculo como se muestra en la figura, la circunferencia y el área del círculo se calcularía usando una herramienta llamada programa.</li>
            <li>En este caso, la fórmula de la izquierda puede ser utilizada.</li>
            <li>La fórmula muestra que la circunferencia y el área del círculo depende del radio.</li>
            <li>A partir de esta misión, usted aprenderá cómo usar variables y cómo resolver problemas usando operadores.</li>
        </ul>
      </td>
  </tr>
</table>

### <FONT COLOR="blue"> 2. Solución </FONT>

#### **2.1. Python contribuyó en gran manera a la observación de la sombra de agujeros negros**

<table style="width:100%; font-size:150%">
  <tr>
    <td rowspan="2"><img style="float: left;" src="https://static.projects.iq.harvard.edu/files/styles/os_files_xlarge/public/eht/files/20190410-78m-800x466.png?m=1554877319&itok=p3yfzMuz" width=500></td>
    <td>
        <ul>
        <li>En abril 10 de 2019, la primera observación de un agujero negro de la humanidad fue exitosa. El agujero negro en sí mismo es invisible y de su sombra se generó una imagen. </li>
        <li>Para esto, los científicos usaron tecnología para obtener imágenes del agujero negro combinando imágenes obtenidas de radiotelescopios en varios lugares en el planeta tierra.</li>
        <li>En otras palabras, la tierra misma fue usada como un gran telescopio, para lo cual 5 Petabytes de información fueron recolectados, muchas herramientas fueron usadas, y la librería Numpy de Python jugó un rol importante.</li>
        </ul>
    </td>
  </tr>
  <tr>
      <td>
          <img style="float: left;"<img src="https://ambg0231.github.io/unidad3_/WorldTeles.png"  width= "800px" />
      </td>
  </tr>
</table>

#### **2.2. Python para resolver la crisis climática global**


<table style='font-size:150%; width:100%'>
  <tr>
      <td>
          <img style="float: left;"<img src="https://ambg0231.github.io/unidad3_/Forecast.png"  width= "800px" />
      </td>
      <td>
        <ul>
            <li>Python se ha convertido en una herramienta esencial en muchos campos, incluyendo la biología, física, y astronomía, así como en las ciencias computacionales y de datos.</li>
            <li>Forecastwatch.com, un servicio que predice el clima mundial, usa Python para procesar la información recolectada de satélites y sensores del clima para hacer pronósticos del clima.</li>
        </ul>
      </td>
  </tr>
</table>

https://forecastwatch.com

### <FONT COLOR="blue"> 3. Misión </FONT>
#### **3.1. Imprimir la circunferencia y el área de un círculo**

| En esta misión, obtenga una entrada de usuario del radio de un círculo.

| Calcule la circunferencia y el área del círculo usando los operadores que aprenderemos en esta unidad.

| Imprima en la pantalla la circunferencia y el área del círculo calculada.

| Ingrese para el radio varios números como 4, 5, 6, 10, 20, 33.

#### **3.2. ¿Cómo se imprime el área y la circunferencia del círculo?**

In [8]:
PI = 3.14
radius = float(input ('Enter the radius :'))
circum = PI * radius * radius
area = 2.0 * PI * radius
print ('Radius of the circle', radius)
print('Area of the circle', circum)
print ('Circumference of the circle', area)

Enter the radius : 6


Radius of the circle 6.0
Area of the circle 113.03999999999999
Circumference of the circle 37.68


### **3.3. Plan de programación**

![Images.png](https://ambg0231.github.io/unidad3_/Charts.png)

#### **3.4. Imprimir el área y la circunferencia del círculo: código final**

In [9]:
PI = 3.14
radius = float(input ('Enter the radius :'))
circum = PI * radius * radius
area = 2.0 * PI * radius
print ('Radius of the circle', radius)
print('Area of the circle', circum)
print ('Circumference of the circle', area)

Enter the radius : 11


Radius of the circle 11.0
Area of the circle 379.94
Circumference of the circle 69.08


## <FONT COLOR="blue"> Concepto Clave </FONT>
### <FONT COLOR="blue"> Tipos de datos numéricos, aritmética y división </FONT>
#### **1.1. Computación usando información numérica**

<table style='font-size:150%; width:100%'>
  <tr>
      <td>
          <img style="float: left;"<img src="https://ambg0231.github.io/unidad3_/Numbers.jpg"  width= "500px" />
      </td>
      <td>
        <ul>
            <li>Python expresa datos numéricos a través de números.</li>
            <li>Se proveen varios operadores matemáticos para estos datos numéricos.</li>
            <li>Se proveen varios operadores matemáticos para estos datos numéricos.</li>
        </ul>
        <img style="float: left;"<img src="https://ambg0231.github.io/unidad3_/Ex.png"  width= "1000px" />
      </td>
  </tr>
</table>

Aprendamos los términos para una operación aritmética. Los valores de datos o variables sujetos a la operación son llamados operandos. El sujeto de la operación es llamado el operador. Tome como ejemplo de una operación de 100 + 200

<img src="https://ambg0231.github.io/unidad3_/Calculator.png"  width= "1000px" />


#### **1.2. Operadores**

La tabla de abajo explica los operadores de Python y el funcionamiento del operador.

<img src="https://ambg0231.github.io/unidad3_/OpTable.png"  width= "1000px" />

#### **1.3. Código de ejemplo usando operadores**

Ejecute una operación usando tipos de datos numéricos. La operación puede ser ejecutada usando datos numéricos y operadores +, -, *, /, //, %

In [10]:
100 + 20 # Find addition of numerical value of 100 and 20

120

In [11]:
100 * 20 # Find multiplication of numerical value of 100 and 20

2000

In [12]:
100 - 20 # Find subtraction of numerical value of 100 and 20

80

In [13]:
100 / 20 # Divide numerical value of 100 with 20

5.0

In [14]:
100 // 20 # Divide numerical value of 100 with 20 and find quotient

5

In [15]:
100 % 20 # Divide numerical value of 100 with 20 and find quotient

0

### <FONT COLOR="blue"> 2. Datos del tipo Objeto, prioridad de operadores, y conversión de tipos de datos </FONT>

#### **2.1. Clasificación de tipos de datos**

En Python, la discriminación de tipos de datos es muy importante.

Los caracteres y los números tiene la misma forma, pero sus propiedades pueden ser muy diferentes.

<span style="color:blue; font-size:200%">**Ex**</span> El número 10 Puede ser calculado. Sin embargo, la letra "10" son imposibles de calcular

Dependiendo del tipo de dato, los operadores disponibles son diferentes. En el siguiente caso, un error gramatical ocurre porque se divide los caracteres "10" entre 2.

<img src="https://ambg0231.github.io/unidad3_/Error1.png"  width= "1000px" />

En Python, el operador tiene diferentes aplicaciones dependiendo del tipo de datos.

El resultado de la operación de multiplicación (*) de los números 100 por 2 is 200.

Sin embargo, el resultado de la operación de multiplicación (*) de las letras '100' por 2 is '100100'. En otras palabras el resultado de la operación de multiplicación (*) de las letras 'Hello' por 2 es 'HelloHello', pero la operación de multiplicación en las letras es diferente a la multiplicación de valores numéricos.

In [16]:
100 * 2 # Find multiplication of numerical value 100 and 2

200

In [17]:
'Hello' * 2 # Find multiplication of string ‘Hello’ and 2

'HelloHello'

In [18]:
'100' * 2 # Find multiplication of string ‘100’ and 2

'100100'

#### **2.2. Tipos de datos básicos en Python**

Los tipos de datos básicos y ejemplos en Python que son ampliamente usados, son los siguientes.

<img src="https://ambg0231.github.io/unidad3_/TiposDatos.png"  width= "1000px" />

Aquí están las estructuras de datos importantes de Python y los tipos de variables. Estos serán usados para codificar.

<img src="https://ambg0231.github.io/unidad3_/DatStruct.png"  width= "1000px" />

#### **2.3. ¿Cómo distinguir el tipo de datos?**

Los términos refiriéndose a valores numéricos, funciones y clases, los cuales son componentes del programa Python, son objetos.

El rol del operador difiere dependiendo si el tipo de datos del objeto es entero o string (caracteres)

La función Type devuelve el tipo de un objeto. Verifiquemos el tipo de datos usando el código de abajo.

<img src="https://ambg0231.github.io/unidad3_/TypeFunc.png"  width= "1000px" />

In [20]:
type(100) # Le muestra el tipo de datos de 100

int

In [21]:
type(100.0) # Le muestra el tipo de datos de 100

float

In [22]:
type('100') # Le muestra el tipo de datos de 100

str

In [23]:
type(10 + 3j) # Los tipos de datos complejos en la forma de un número real + un número imaginario como 10 + 3j

complex

#### **2.4. Precedencia de los operadores**

Si los operadores de adición y multiplicación aparecen en una sentencia, la operación de multiplicación se ejecuta primero.

Cuando aparece un paréntesis en una sentencia, los operadores en paréntesis se tratan primero.

Los cálculos en la sentencia varían de acuerdo al orden de procesamiento de los operadores.

In [24]:
10 + 20 * 30 # La multiplicación se calcula primero

610

In [26]:
(10 + 20) * 30 # El operador entre paréntesis se calcula primero

900

#### **2.5. Diagrama de precedencia de operadores**

Veamos las precedencias de los operadores de Python. Estos son los contenidos a aprender en esta lección.

<img src="https://ambg0231.github.io/unidad3_/OperTable.png"  width= "1000px" />

#### **2.6. Conversión del tipo de dato**

Los datos en Python se pueden calcular convirtiendo su tipo. Los caracteres '10' se pueden convertir en tipo entero con int('10'). Por lo tanto int('10')/2 se ejecuta sin errores.

Para cambiar el número 10 de dato en Python al carácter '10', se usa str(10). Este carácter convertido se puede escribir con otra cadena de caracteres como una operación de adición.

In [28]:
int('10') / 2 # Convertir el carácter '10' en un entero y dividirlo entre 2

5.0

In [29]:
'I like number ' + str(10) # COnvertir el número 10 en carácter y usar el operador +

'I like number 10'

In [30]:
a = None
type(a)

NoneType

<img src="https://ambg0231.github.io/unidad3_/OMS1.png"  width= "1000px" />

### <FONT COLOR="blue">3. Aplicación de String </FONT>

#### **3.1. Veamos el *string***

EN Python, el string str es muy apropiado para mostrar información en un software.

Todos los archivos str pueden ser convertidos a texto. Sin embargo, dado que es un texto, es imposible aplicarle operaciones aritméticas.

<span style="color:blue; font-size:200%">**Ex**</span> El nombre del usuario, ID, contraseña, etc., se puede expresar como un string.

Los strings de Python son formas en las cuales cada letra está conectada.

Las operaciones de adición y multiplicación son posibles en los strings de Python.

<img src="https://ambg0231.github.io/unidad3_/String.png"  width= "1000px" />

#### **3.2. El operador string (+),(*)**

Veamos las operaciones + y * disponibles en string. La operación + del string conecta dos strings. La operación * del string debe ser un entero. Es decir, el string se genera repitiendo el entero n.

In [35]:
'I'+ ' Love' + ' Python!' # Un string se puede sumar a un string.

'I Love Python!'

In [36]:
'Python ' * 10 # Genera un string de "Python repetido 10 veces"

'Python Python Python Python Python Python Python Python Python Python '

In [37]:
'*' * 50 # Genera un string de '*' repetido 50 veces

'**************************************************'

In [38]:
str(100) * 10 # El string '100' repetido 10 veces

'100100100100100100100100100100'

||| La operación * del string debe ser un entero. De lo contrario ocurrirá un error|||

In [39]:
'*' * 3.1

TypeError: can't multiply sequence by non-int of type 'float'

<img src="https://ambg0231.github.io/unidad3_/OMS2.png"  width= "1000px" />

## <FONT COLOR="blue"> Codificación en papel </FONT>

<table style="font-size:24px">
  <tr>
    <td>
        Trate de entender completamente los conceptos básicos antes de moverse al siguiente paso.
    </td>
  </tr>
  <tr>
    <td>
        La falta de entendimiento de los conceptos básicos incrementará su carga en el aprendizaje de este curso, lo que puede hacer que lo repruebe.
    </td>
  </tr>
  <tr>
    <td>
        Puede que sea difícil ahora, pero para completar con éxito este curso le sugerimos entender completamente los conceptos y luego pasar a el siguiente paso.
    </td>
  </tr>
</table>

### <FONT COLOR="blue">**Q1.** </FONT>

Use la función int para retornar el número 100 sumando el carácter "50" al número 50. Adicionalmente, use la función str para retornar el resultado de la adición "5050".

<table style='font-size:150%; width:100%'>
  <tr>
      <td>
          <b>Condición para la ejecución</b>
      </td>
      <td>
          La primera línea retorna el número 100, y la segunda línea retorna los caracteres "5050"
      </td>
  </tr>
  <tr>
      <td>
          <b>Tiempo</b>
      </td>
      <td>
          5 Minutos
      </td>
  </tr>
</table>

| Escriba el código y el resultado de salida esperado en una hoja

### <FONT COLOR="blue">**Q2.** </FONT>

Se proveen una letra '1' y tres '0'. Use estas para construir el número 1000. Aquí solo se permiten operaciones de adición entre strings, y la función int() puede ser usada sólo una vez.

<table style='font-size:150%; width:100%'>
  <tr>
      <td>
          <b>Condición para la ejecución</b>
      </td>
      <td>
          Retorne el número '1000'
      </td>
  </tr>
  <tr>
      <td>
          <b>Tiempo</b>
      </td>
      <td>
          5 Minutos
      </td>
  </tr>
</table>

| Escriba el código y el resultado de salida esperado en una hoja

## <FONT COLOR="blue"> Vamos a codificar </FONT>

### <FONT COLOR="blue"> 1. Tipos de datos numéricos, aritmética y división</FONT>

#### **1.1. Operaciones aritméticas**



<table style='font-size:150%; width:100%'>
  <tr>
      <td>
          <img src="https://ambg0231.github.io/unidad3_/WoodenCube.png" controls="controls" style="width: 500px;">
      </td>
      <td>
        <ul>
            <li>Implemente varias operaciones matemáticas en Python, incluyendo adición, sustracción, multiplicación, división y raíz cuadrada.</li>
        </ul>
      </td>
  </tr>
</table>

Realice cálculos aritméticos codificando

In [40]:
1 + 5

6

In [41]:
1 - 8

-7

In [42]:
4 * 5

20

In [43]:
5 // 3

1

**Línea 1**

* Calcule la adición

**Línea 2**

* Calcule la sustracción

**Línea 3**

* Calcule la multiplicación

**Línea 2**

* División piso (cociente): retorna 1, que es el cociente de dividir 5 entre 3.

In [44]:
2 ** 10

1024

In [45]:
9 % 5

4

**Línea 1 ~ 2**

* 1: Se ejecuta la potencia 2<sup>10</sup> para devolver 1024

* 2: Se llama módulo. Devuelve el residuo de 9 dividido entre 5 el cual es 4.

In [47]:
1 + 2 * 5

11

In [49]:
(1 + 2) * 5

15

**Línea 1 ~ 2**

* 1: Las operaciones numéricas tiene la misma precedencia que las operaciones aritméticas generales. En otras palabras, la multiplicación y la división se ejecutan antes que la adición.

* 2: Si hay un paréntesis, se calcula este primero.

In [50]:
(3 + 4j) + (5 + 2j) # Adición de números complejos

(8+6j)

In [51]:
(3 + 4j) - (5 + 2j) # Sustracción de números complejos

(-2+2j)

In [52]:
(3 + 4j) * (5 + 2j) # Multiplicación de números complejos

(7+26j)

In [53]:
(3 + 4j) / (5 + 2j) # División de números complejos

(0.793103448275862+0.48275862068965514j)

**Línea 1 ~ 4**

* Adición, sustracción, multiplicación y división de números complejos.


### <FONT COLOR="blue"> 2. Tipos de datos de objetos, prioridades de operadores y conversión de tipos de datos </FONT>
#### **2.1. Identificar y convertir tipos de datos**

La identificación y conversión de un cierto valor de un tipo de dato es un tema importante.


_______________ int("10")  <span style="color:blue"># Convierte el string "10" en el entero 10</span> __________________


El tipo de dato es crítico porque el operador a ser usado varía dependiendo el tipo de dato.

In [54]:
type(10)

int

In [55]:
type("10")

str

In [56]:
type(10.5)

float

**Línea 1**

* Cuando un número (entero) es una entrada, se marca como un int.

**Línea 2**

* Cuando un string es una entrada, se marca como str(string)

**Línea 3**

* Cuando un número (número real) con puntos decimales es una entrada, se marca como float.

<img src="https://ambg0231.github.io/unidad3_/OMS3.png"  width= "1000px" />

#### **2.2. ¿Cómo convertir un tipo de dato?**

In [58]:
int("50")

50

In [59]:
int("50") + 20

70

In [60]:
"10" + str(10)

'1010'

In [61]:
float("10.5")

10.5

**Línea 1 ~ 2**

* 1: Use la función int para remplazar letras por números.

* 2: Dado que el string "50" se ha cambiado a un entero a través de int("50"), la operación + 20 se ejecuta.

**Línea 3 ~ 4**

* 3: Para remplazar números por letras, use la función str.

* 4: Use la función float para remplazar la letra por un número real.

### <FONT COLOR="blue">3. Strings y Operadores</FONT>
#### **3.1. Operadores de adición y multiplicación de strings**

El string puede usar operadores como adición y multiplicación. La sustracción y división no pueden usarse con estos.

In [62]:
"Helo" + "World"

'HeloWorld'

In [64]:
"text" + "repeat" * 10

'textrepeatrepeatrepeatrepeatrepeatrepeatrepeatrepeatrepeatrepeat'

**Línea 1**

* Un string no se puede sustraer, pero se puede adicionar. Dos strings se pueden adicionar para conectarlos.

**Línea 2**

* Cuando un string se multiplica por un entero, el string se repite. Aquí, 'repeat' se repite 10 veces.

* Sin embargo, la división y la sustracción no funcionan.

### **3.2. Comillas simples y dobles**

Cuando se imprime un string, se obtiene el mismo resultado si se usa comillas simples como 'Hello Python!!' o dobles como "Hello Python!!".

In [65]:
print("Hello World")
print('Hello World')

Hello World
Hello World


**Línea 1 ~ 2**

* Los strings " ~~ " y ' ~~ ' muestran el mismo resultado.

In [66]:
print('I am', 23, 'years old')
print('I', 'really ' * 3, 'love', 'Python!')

I am 23 years old
I really really really  love Python!


**Línea 1 ~ 2**

* Se imprimieron valores numéricos y strings usando comas.

* La coma automáticamente añade un espacio cuando retorna la salida.

### **3.3. Secuencia de escape**

La secuencia de escape \ se puede usar para retornar comillas simples o dobles en la pantalla.

In [67]:
print("\"double quotation marks\"")

"double quotation marks"


In [68]:
print("\'double quotation marks\'")

'double quotation marks'


**Línea 1 ~ 2**

* Los strings deben estar encerrados en comillas simples o dobles.

* Por ello es difícil indicar comillas. En este caso, usted puede usar \\.

* Una secuencia de escape se refiere a un caracter especial usado para controlar un computador, y dado que una comilla simple o doble se define para definir un string, se retorna usando el caracter de control \\ para retornarlo.

La secuencia de escape: \n se usa para pasar a la siguiente línea.

In [69]:
print("Line\nbreak")

Line
break


**Línea 1**

* <Barra diagonal invertida + caracter> se usa para mostrar caracteres especiales como un cambio de línea. \n es un caracter para campiar a la siguiente línea, y \t es un caracter de tabulación.

La secuencia de escape: \t se usa cuando se quiere ingresar una tabulación.

In [71]:
print("Enter a \t tab in the middle of the sentence.")

Enter a 	 tab in the middle of the sentence.


**Línea 1**

* Cuando se quiera ingresar una tabulación en medio de una sentencia, ingrese \t en la ubicación correspondiente.

In [72]:
print("\tEnter a tab at first.")

	Enter a tab at first.


**Línea 1**

* Cuando se quiere ingresar una tabulación al inicio de una sentencia, ingrese \t como el primer caracter de salida.

#### **3.5. Indexación de strings**

Indexación de strings: Se puede usar para obtener <span style="color:blue">solo parte de una larga cadena de caracteres</span>. Use los brackets[] y números.

In [74]:
"hello"[0]

'h'

In [75]:
"hello"[2]

'l'

**Línea 1**

* La indexación de strings empieza por 0.

* Es decir, cuando se indexa 0, h, la primera letra de "hello" se indexa.

**Línea 2**

* Cuando se usa 2 como índice en la indexación de strings, usted puede ver que la tercera letra "l" se indexa.

#### **3.6. Indexación negativa de strings**

Indexación de strings: Se puede usar para obtener <span style="color:blue">solo parte de una larga cadena de caracteres</span>. Use los brackets[] y números.

In [76]:
"hello"[-1]

'o'

**Línea 1**

* Python soporta indexación negativa y cuando se indexa -1, se indexa el último caracter.

In [77]:
"hello"[-5]

'h'

**Línea 1**

* Python soporta indexación negativa y cuando se indexa -5, se indexa el primer caracter.

<img src="https://ambg0231.github.io/unidad3_/Attention.png"  width= "1000px" />

**Línea 1**

* El string 'hello', tiene 5 caracteres. Por lo tanto el rango de indexación es de 0 a 4. Precaución: un ídice fuera de rango genera un error.

## <FONT COLOR="blue">Programación por parejas </FONT>

### <FONT COLOR="blue">Práctica de la programación en parejas </FONT>


**Directriz, mecanismos y plan de contingencia**

 Preparar la programación por parejas implica establecer pautas y mecanismos para ayudar a los alumnos a emparejarse correctamente y mantenerlos emparejados.  Por ejemplo, los estudiantes deben turnarse para "conducir el mouse".  La preparación efectiva requiere planes de contingencia en caso de que uno de los socios se ausente o decida no participar por una u otra razón.  En estos casos, es importante dejar claro que el alumno activo no será sancionado porque el emparejamiento no funcionó bien.
 
**Emparejar habilidades similares, no necesariamente iguales, como socios**

 La programación en parejas puede ser efectiva cuando los estudiantes con habilidades similares, aunque no necesariamente iguales, se emparejan como compañeros.  El emparejamiento de estudiantes que no coinciden a menudo puede conducir a una participación desequilibrada.  Los maestros deben enfatizar que la programación en parejas no es una estrategia de "divide y vencerás", sino más bien un verdadero esfuerzo de colaboración en cada esfuerzo por todo el proyecto.  Los maestros deben evitar emparejar a estudiantes muy débiles con estudiantes muy fuertes.

**Motivar a los estudiantes ofreciendo incentivos adicionales**

 Ofrecer incentivos adicionales puede ayudar a motivar a los estudiantes a emparejarse, especialmente con estudiantes avanzados.  A algunos maestros les ha resultado útil pedirles a los estudiantes que se emparejen solo para una o dos tareas.

**Evitar trampas en la colaboración**

 El desafío para el maestro es encontrar formas de evaluar los resultados individuales, mientras aprovecha los beneficios de la colaboración.  ¿Cómo saber si un estudiante aprendió o hizo trampa?  Los expertos recomiendan revisar el diseño y la evaluación del curso, así como discutir explícita y concretamente con los estudiantes los comportamientos que se interpretarán como trampa.  Los expertos alientan a los maestros a hacer que las tareas sean significativas para los estudiantes y a explicar el valor de lo que los estudiantes aprenderán al completarlas.

**Entorno de aprendizaje colaborativo**

 Un entorno de aprendizaje colaborativo ocurre cada vez que un instructor requiere que los estudiantes trabajen juntos en actividades de aprendizaje.  Los entornos de aprendizaje colaborativo pueden implicar actividades tanto formales como informales y pueden incluir o no una evaluación directa.  Por ejemplo, parejas de estudiantes trabajan en asignaciones de programación; pequeños grupos de estudiantes discuten posibles respuestas a la pregunta de un profesor durante la lección; y los estudiantes trabajan juntos fuera de clase para aprender nuevos conceptos.  El aprendizaje colaborativo es distinto de los proyectos en los que los estudiantes "dividen y vencerán".  Cuando los estudiantes dividen el trabajo, cada uno es responsable de solo una parte de la resolución del problema y hay oportunidades muy limitadas para resolver los problemas con los demás.  En entornos colaborativos, los estudiantes participan en conversaciones intelectuales entre ellos.



### <FONT COLOR="blue">**Q1.** </FONT>

n! se define como n * (n-1) * (n-2) * (n-3) *...* 2 * 1. Encuentre 5! y 10! usando enteros y el operador * e imprimalos como sigue.

### <span style="color:blue">Ejemplo de resultado</span>

Calcule el factorial.

5! = 120

Calcule el factorial.

10! = 3628800